Give rules access to settings (#279)
remarkjs/remark-lint: readme.md

<a name="configuring-remark-lint"></a>

Just like other remark plugins, there are a couple ways to configure rules.
They all follow a standard approach:

*   Pass a severity, options, or both to the rules themselves.
    This affects each rule individually and takes highest precedence.

> *   Some rules correspond to settings that are shared among all remark plugins --- particularly [`remark-stringify`](https://github.com/remarkjs/remark/tree/main/packages/remark-stringify#options).

Author: - [x] I've added shared settings docs to the Configure section.

  // Use `'error'` or `2` to treat messages as exceptions:
  .use(remarkLintFinalNewline, ['error'])
  .use(remarkLintFinalNewline, [2])

>   // Some rules correspond to `remark-stringify` settings which they obey by default:
>   .use(remarkLintUnorderedListMarkerStyle).use({settings: {bullet: '*'}})
>   // Some accept options, and what they exactly accept is different for

Reviewer: ```suggestion
  // Some rules respond to shared settings that `remark-stringify` also understands:
  .use({settings: {bullet: '*'}})
  .use(remarkLintUnorderedListMarkerStyle)
  // Some rules accept options, and what they exactly accept is different for
```

(I do think that this new example should be *added*, after the existing examples, though)

| [`settings.quote`](https://github.com/remarkjs/remark/tree/main/packages/remark-stringify#optionsquote) | [`remark-lint-link-title-style`](https://github.com/remarkjs/remark-lint/tree/main/packages/remark-lint-link-title-style) |
| [`settings.rule`](https://github.com/remarkjs/remark/tree/main/packages/remark-stringify#optionsrule) | [`remark-lint-rule-style`](https://github.com/remarkjs/remark-lint/tree/main/packages/remark-lint-rule-style) |
| [`settings.ruleRepetition`](https://github.com/remarkjs/remark/tree/main/packages/remark-stringify#optionsrulerepetition) | [`remark-lint-rule-style`](https://github.com/remarkjs/remark-lint/tree/main/packages/remark-lint-rule-style) |
| [`settings.ruleSpaces`](https://github.com/remarkjs/remark/tree/main/packages/remark-stringify#optionsrulespaces) | [`remark-lint-rule-style`](https://github.com/remarkjs/remark-lint/tree/main/packages/remark-lint-rule-style) |
| [`settings.setext`](https://github.com/remarkjs/remark/tree/main/packages/remark-stringify#optionssetext) | [`remark-lint-heading-style`](https://github.com/remarkjs/remark-lint/tree/main/packages/remark-lint-heading-style) |

> | [`settings.strong`](https://github.com/remarkjs/remark/tree/main/packages/remark-stringify#optionsstrong) | [`remark-lint-strong-marker`](https://github.com/remarkjs/remark-lint/tree/main/packages/remark-lint-strong-marker) |

Author: - [x] I've added a table to the Configure section, listing the settings that also affect lint rules.

    .use(remarkPresetLintConsistent)
    // Few recommended rules.
    .use(remarkPresetLintRecommended)

>     // `remark-lint-list-item-indent` is configured with `tab-size` in the

Reviewer: I think this change should be reverted. See: https://github.com/remarkjs/remark-lint/pull/281#issuecomment-1050911561.

    .use(remarkPresetLintConsistent)
    // Few recommended rules.
    .use(remarkPresetLintRecommended)
    .use({
      // `remark-lint-list-item-indent` is configured with `tab` in the
      // recommended preset, but if we’d prefer something else, it can be
      // reconfigured:
      settings: {listItemIndent: 'one'}

>     })

Reviewer: I believe the prose here does not explain what’s happening correctly. The rule is not configured in the preset. The preset defines a shared setting that the rule and remark-stringify respond to.

Author: - [x] I've edited the prose:
> `settings.listItemIndent` is set to `tab` in the recommended preset, but you can change it if you’d prefer something else:

remark lint rules *check* markdown.
[`remark-stringify`][remark-stringify] (used in remark) *formats* markdown.
When you configure lint rules and use remark to format markdown, you must

> manually synchronize their configuration:

Reviewer: I think this statement is still correct. There is a new way to solve it though. That can be added as a new example. It is still important to call out the problem though.

```js
  /* … */
  "remarkConfig": {
    "settings": {
      // `remark-lint-list-item-indent` is configured with `tab` in the
      // recommended preset, but if we’d prefer something else, it can be
      // reconfigured:
      "listItemIndent": "one"
    },
    "plugins": [
      "remark-preset-lint-consistent", // Check that markdown is consistent.

>       "remark-preset-lint-recommended" // Few recommended rules.

Reviewer: See above (I think the wording on what’s happening can improve)

  "remarkConfig": {
    "settings": {
      "emphasis": "*",
      "listItemIndent": "one",
      "strong": "*"
    },
    "plugins": [
      "remark-preset-lint-consistent",
      "remark-preset-lint-recommended"
    ]
  },
  /* … */
```

`settings` configures both

> [`remark-stringify`][remark-stringify] and the corresponding rules, and explicitly prefers asterisks

Reviewer: This is explained better than the previous ones IMO, but I still see room for improvement. E.g., only asterisks is mentioned but not `one`?

Author: - [x] I've removed `listItemIndent` from this example: It isn't mentioned in the current prose, either.
